### ISSUE TITLE
How to disable service binding from Spring Configuration

### ISSUE THREAD

**Issue by mbwhelan** [pivotal-cf/java-cfenv#93]
I am having issues being able to disable service bindings from being loaded.  I have two rabbitmq services bound to my application: rabbitmq, rabbitmq-spring-cloud-bus.  I can only get it to work if I use a system property like, `-Dcfenv.service.rabbitmq.enabled=false`, via JAVA_OPTS   It doesn't not seem to work if set in Spring Cloud configuration or a local application.yml file. Is this as designed?

If i set `cfenv.service.rabbitmq.enabled: false` in application.yml, I get the following error:
```
2020-06-08T12:07:35.489-05:00 [APP/PROC/WEB/0] [OUT] 17:07:35.489 [main] ERROR org.springframework.boot.SpringApplication - Application run failed
2020-06-08T12:07:35.489-05:00 [APP/PROC/WEB/0] [OUT] java.lang.IllegalArgumentException: No unique Rabbit service found. Found service names [rabbitmq, rabbitmq-spring-cloud-bus]
2020-06-08T12:07:35.489-05:00 [APP/PROC/WEB/0] [OUT] at io.pivotal.cfenv.spring.boot.CfEnvironmentPostProcessor.throwExceptionIfMultipleMatches(CfEnvironmentPostProcessor.java:154)
2020-06-08T12:07:35.489-05:00 [APP/PROC/WEB/0] [OUT] at io.pivotal.cfenv.spring.boot.CfEnvironmentPostProcessor.postProcessEnvironment(CfEnvironmentPostProcessor.java:94)
2020-06-08T12:07:35.489-05:00 [APP/PROC/WEB/0] [OUT] at org.springframework.boot.context.config.ConfigFileApplicationListener.onApplicationEnvironmentPreparedEvent(ConfigFileApplicationListener.java:200)
2020-06-08T12:07:35.489-05:00 [APP/PROC/WEB/0] [OUT] at org.springframework.boot.context.config.ConfigFileApplicationListener.onApplicationEvent(ConfigFileApplicationListener.java:188)
2020-06-08T12:07:35.489-05:00 [APP/PROC/WEB/0] [OUT] at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:172)
2020-06-08T12:07:35.489-05:00 [APP/PROC/WEB/0] [OUT] at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:165)
2020-06-08T12:07:35.489-05:00 [APP/PROC/WEB/0] [OUT] at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:139)
2020-06-08T12:07:35.489-05:00 [APP/PROC/WEB/0] [OUT] at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:127)
2020-06-08T12:07:35.489-05:00 [APP/PROC/WEB/0] [OUT] at org.springframework.boot.context.event.EventPublishingRunListener.environmentPrepared(EventPublishingRunListener.java:80)
2020-06-08T12:07:35.489-05:00 [APP/PROC/WEB/0] [OUT] at org.springframework.boot.SpringApplicationRunListeners.environmentPrepared(SpringApplicationRunListeners.java:53)
2020-06-08T12:07:35.489-05:00 [APP/PROC/WEB/0] [OUT] at org.springframework.boot.SpringApplication.prepareEnvironment(SpringApplication.java:345)
2020-06-08T12:07:35.489-05:00 [APP/PROC/WEB/0] [OUT] at org.springframework.boot.SpringApplication.run(SpringApplication.java:308)
```

Example VCAP_SERVICES
```json
{
  "p.rabbitmq": [
    {
      "label": "p.rabbitmq",
      "provider": null,
      "plan": "small",
      "name": "rabbitmq",
      "tags": [
        "rabbitmq"
      ],
      "instance_name": "rabbitmq",
      "binding_name": null,
      "credentials": {
        "uri": "amqp://username:password@rabbitmq/guid",
        "http_api_uri": "http://username:password@rabbitmq/api"
      },
      "syslog_drain_url": null,
      "volume_mounts": []
    },
    {
      "label": "p.rabbitmq",
      "provider": null,
      "plan": "small",
      "name": "rabbitmq-spring-cloud-bus",
      "tags": [
        "rabbitmq"
      ],
      "instance_name": "rabbitmq-spring-cloud-bus",
      "binding_name": null,
      "credentials": {
        "uri": "amqp://username:password@springcloudbus/guid",
        "http_api_uri": "http://username:password@springcloudbus/api"
      },
      "syslog_drain_url": null,
      "volume_mounts": []
    }
  ]
}
```

**Comment by dyroberts** [comment]
Considering the Is Enabled feature was sitting on a branch until recently, I'm going to make the call that we should make it work from application properties. So, the properties set by java-cfenv won't be accessible in placeholders in application properties.

**Comment by dyroberts** [comment]
Fixed in 6850d22

**Comment by dyroberts** [comment]
@mbwhelan Regarding your comment on issue 94 (https://github.com/pivotal-cf/java-cfenv/issues/94#issuecomment-643290373), would you say that it would be an improvement to be able to disable an entire service type from being autoconfigured? So that you can configure multiple service instances of that type?

**Comment by joaodias14** [comment]
@dyroberts, that would be an improvement, but isn't it possible to simply ignore multiple bindings by default instead of relying on yet another configuration?

**Comment by mbwhelan** [comment]
I would prefer if there are multiple bindings, it logs a warning vs throwing an exception and not setting any the spring.{service}.* properties.  That is currently how the jdbc implemenation is working.

**Comment by dyroberts** [comment]
@joaodias14 @mbwhelan Fair enough. Let's do that! I should be able to get this change in and do a release this week.

**Comment by joaodias14** [comment]
Many thanks @dyroberts! If we can help, just let us know.

**Comment by dyroberts** [comment]
@joaodias14 @mbwhelan I've added this change to master and pulled in the java-cfenv into the spring-cloud-services starters snapshot as well. Would one or both of you mind giving it a try with a snapshot build and let me know if it's working as you'd expect? All my tests are green and I'm ready to release it if so.

**Comment by joaodias14** [comment]
Hi @dyroberts, many thanks for that. I will try to give it a try tomorrow.

**Comment by joaodias14** [comment]
Hi @dyroberts, is there any public repository where this snapshot build is available?

**Comment by joaodias14** [comment]
Nevermind, just found it: https://repo.spring.io/snapshot, right?

**Comment by joaodias14** [comment]
Hi @dyroberts all good now. Now thrown exception. Now we just need to check how to get the multiple instances data but that is another topic. Many thanks for this.

**Comment by joaodias14** [comment]
Hi @dyroberts, I just noticed that it is also not possible to get the services' details by tag when we have multiple instances with the same tag. An exception is thrown. Should the list of available services be returned instead? In my opinion, this assumption that multiple instances should never be in place and an exception should be thrown in such cases will only create issues to a lot people migrating to Java CFEnv.

**Comment by dyroberts** [comment]
Thanks @joaodias14. I'll look into it today!

**Comment by dyroberts** [comment]
@joaodias14 You'll notice that there are two find-by-tag methods on the core `CfEnv` class: `findServiceByTag` and the plural `findServicesByTag`. I think you just need the second one.

**Comment by joaodias14** [comment]
Ups, that is in fact true. Sorry, my mistake. Thanks for the heads up!

**Comment by joaodias14** [comment]
Hi @dyroberts, quick question. Do you have your next release planned already? Just so that we know when this will become available in a non-snapshot version.

**Comment by naavo** [comment]
@dyroberts - we are running into the same issue except our use case requires binding to multiple redis instances. Trying to disable a single instance of Redis via spring cloud config property fails, where as if we set the respective property as JVM property / in bootstrap.yml we could get around this. I tried pulling in the latest  v2.2.0 (cfjava-env-boot) which got released a day ago and it seems to have the same issue. 

I see there are unit test cases around this in the library as per which it should work - i think it boils down to the order in which the property is being evaluated
